src/main/java/net/snowflake/ingest/utils/HttpUtil.java

```diff
@@ -174,4 +225,82 @@ private static String getRequestUriFromContext(final HttpContext httpContext) {
     HttpRequest httpRequest = clientContext.getRequest();
     return httpRequest.getRequestLine().getUri();
   }
+
+  /** Thread to monitor expired and idle connection, if found clear it and return it back to pool */
+  private static class IdleConnectionMonitorThread extends Thread {
+
+    private final PoolingHttpClientConnectionManager connectionManager;
+    private volatile boolean shutdown;
+
+    public IdleConnectionMonitorThread(PoolingHttpClientConnectionManager connectionManager) {
+      super();
+      this.connectionManager = connectionManager;
+    }
+
+    @Override
+    public void run() {
+      try {
+        LOGGER.debug("Starting Idle Connection Monitor Thread ");
+        synchronized (this) {
+          while (!shutdown) {
+            wait(MONITOR_THREAD_INTERVAL_MS);
+
+            StringBuilder sb = new StringBuilder();
+
+            sb.append(
+                createPoolStatsInfo("Total Pool Stats = ", connectionManager.getTotalStats()));
+            Set<HttpRoute> routes = connectionManager.getRoutes();
+
+            if (routes != null) {
+              for (HttpRoute route : routes) {
+                sb.append(createPoolStatsForRoute(connectionManager, route));
+              }
+            }
+
+            LOGGER.debug("[IdleConnectionMonitorThread] Pool Stats:\n" + sb);
+
+            // Close expired connections
+            connectionManager.closeExpiredConnections();
+            // Optionally, close connections
+            // that have been idle longer than 30 sec
+            connectionManager.closeIdleConnections(
+                DEFAULT_IDLE_CONNECTION_TIMEOUT_SECONDS, TimeUnit.SECONDS);
+          }
+        }
+      } catch (InterruptedException ex) {
+        LOGGER.debug("Terminating Idle Connection Monitor Thread ");
+      }
+    }
+
+    public void shutdown() {
+      LOGGER.debug("Shutdown Idle Connection Monitor Thread ");
+      shutdown = true;
+      synchronized (this) {
+        notifyAll();
+      }
+    }
+  }
+
+  /** Create Pool stats for a route */
+  private static String createPoolStatsForRoute(
+      PoolingHttpClientConnectionManager connectionManager, HttpRoute route) {
+    PoolStats routeStats = connectionManager.getStats(route);
+    return createPoolStatsInfo(
```

Comment:
why do we need `createPoolStatsInfo` to append two strings? seems unnecessary. I would just do it inline since a StringBuilder will not give you any benefit here.

Reply:
I will need it to print total stats and one per each route in the connection.

Reply:
still, it does not seem like a StringBuilder is useful here with the current usage

Reply:
Yes, removed the usage of SB.